hypothesis-python/src/hypothesis/internal/conjecture/datatree.py

```diff
@@ -67,6 +67,83 @@ class Conclusion:
     interesting_origin = attr.ib()
 
 
+# The number of max children where, beyond this, it is practically impossible
+# for hypothesis to saturate / explore all children nodes in a reasonable time
+# frame. We use this to bail out of expensive max children computations early,
+# where the numbers involved are so large that we know they will be larger than
+# this number.
+#
+# Note that it's ok for us to underestimate the number of max children of a node
+# by using this. We just may think the node is exhausted when in fact it has more
+# possible children to be explored. This has the potential to finish generation
+# early due to exhausting the entire tree, but that is quite unlikely: (1) the
+# number of examples would have to be quite high, and (2) the tree would have to
+# contain only one or two nodes, or generate_novel_prefix would simply switch to
+# exploring another non-exhausted node.
+#
+# Also note that we may sometimes compute max children above this value. In other
+# words, this is *not* a hard maximum on the computed max children. It's the point
+# where further computation is not beneficial - but sometimes doing that computation
+# unconditionally is cheaper than estimating against this value.
+#
+# The one case where this may be detrimental is fuzzing, where the throughput of
+# examples is so high that it really may saturate important nodes. We'll cross
+# that bridge when we come to it.
+MAX_CHILDREN_EFFECTIVELY_INFINITE = 100_000
```

Comment:
number up for consideration. I would be a little uneasy going lower than 25k, which is in the realm of "one person in a thousand with a crazy pseudo-fuzzing workflow might actually hit this".

Reply:
In (1), I'm concerned about the case of saturating the `n` children but continuing to try to draw because we set a max count higher than the actual number of max children.

To restate your cases slightly, we can either (1) overapproximate or (2) underapproximate the number of max children. The current code underapproximates: `count = min(count, EFFECTIVELY_INFINITE)`. This means that we'll consider the tree exhausted before it really is, and will miss some nodes.

In case (1), we overapproximate for large `n`: `count = max(count, 2 ** 64) if count > EFFECTIVELY_INFINITE`. This means that we really will explore every node, but we'll *continue* to try to explore past that point, because we don't think the tree has been saturated yet. I believe that if we do this, we have the potential to infinitely loop when trying to choose a  new child for a node: https://github.com/HypothesisWorks/hypothesis/blob/775ed9125f8429b6a92912c723d0cefee8b664cf/hypothesis-python/src/hypothesis/internal/conjecture/datatree.py#L236-L255

if (a) we've drawn every possible child, but it thinks we haven't because we overapproximated, and (b) every child that we have drawn is exhausted, so it can only try to find a new un-exhausted child. After 1k iterations we'd hit our failsafe and error.

Reply:
Got it, that's a good argument against splitting the concepts.  Let's go with the current proposal then, and as your comment says we can deal with the fuzzing-etc workloads later if they become a problem.